Add a few special disallowed headers

captainbrosset opened the pull request:
This PR should fix #1096.

It adds warnings for the following HTTP response headers:

* Expires
* Host
* P3P
* Pragma
* Via
* X-Frame-Options
* X-UA-Compatible

These were added based on https://www.fastly.com/blog/headers-we-dont-want.

Instead of adding these headers to the existing list of disallowed headers, I added specific checks for these ones. The idea is, this way, we can have specific messages for each of them.

I created a couple of preliminary commits to prepare the code for this, then added the logic for this in a third commit, and finished adding all warnings in a 4th commit.
Finally, I made sure the ignored headers, as configured by users, would also ignore these special headers.

captainbrosset commented:
Note that I did sign the CLA, but with my microsoft email address. I just added it to my github account now. So maybe the licence/cla bot needs to re-run or something.

antross commented:
I'm not sure why the CLA bot didn't pick up the change here, but I can see it worked correctly for your second PR. We can just overrule the bot for this one rather than jump through hoops to figure out how to get it to reset.

antross commented on packages/hint-no-disallowed-headers/src/_locales/en/messages.json:
```diff
@@ -11,6 +11,34 @@
+        "description": "Report message when the response includes the X-Frame-Options header",
+        "message": "The 'X-Frame-Options' header should not be used, the same effect can be achieved with more consistent support using the 'Content-Security-Policy' header."
+    },
+    "disallowedXUACompatibleHeader": {
```
Apologies for missing this in our earlier discussion, but `X-UA-Compatible` overlaps with our recommendations from `hint-highest-available-document-mode` (https://webhint.io/docs/user-guide/hints/hint-highest-available-document-mode/). 

That hint is a touch more nuanced: it recommends removing the meta tag or header if targeting IE11+, but recommends including it if for some crazy reason you're still targeting IE10 or below to ensure you're running in the latest standards mode.

We should probably not include `X-UA-Compatible` as part of this hint to avoid redundant reports.

captainbrosset replied:
Good to know!
I changed the code to not include `X-UA-Compatible`.

antross commented on packages/hint-no-disallowed-headers/src/_locales/en/messages.json:
```diff
@@ -11,6 +11,30 @@
     },
+    "disallowedExpiresHeader": {
+        "description": "Report message when the response includes the Expires header",
+        "message": "The 'Expires' header should not be used, 'Cache-Control' should be preferred."
```
Thanks again for breaking these out! I really appreciate having actionable messages, especially when we're suggesting developers use an alternative (as opposed to simply removing a header).

antross commented on packages/hint-no-disallowed-headers/src/hint.ts:
```diff
@@ -43,75 +94,53 @@ export default class NoDisallowedHeadersHint implements IHint {
-            });
-        };
+        const validateGeneralHeaders = ({ response, resource }: FetchEnd) => {
+            const headers: string[] = includedHeaders(response.headers, generalDisallowedHeaders);
```
You don't actually need to change this since it was pre-existing code you moved, but I wanted to call out we've been slowly moving toward dropping redundant types in our existing code when we make updates. Keeps things a bit easier to read and closer to JS. Probably worth us finding/creating a linting rule at some point to track down the rest of these...
```suggestion
            const headers = includedHeaders(response.headers, generalDisallowedHeaders);
```

captainbrosset replied:
Thanks for letting me know. I'll do this change since it's so simple. And I'll keep that in mind for the future.

antross commented on packages/hint-no-disallowed-headers/src/hint.ts:
```diff
@@ -27,13 +27,64 @@ const debug = d(__filename);
 
     public static readonly meta = meta;
 
+    private serverHeaderContainsTooMuchInformation(serverHeaderValue: string): boolean {
```
Nit: I'm glad to see this moved out of the constructor, but since it doesn't use any instance state can we either make it `static` or move it outside of the class entirely as a helper function (the latter being my personal preference)?

captainbrosset replied:
Ah, thanks, I wasn't sure whether to do this or not. But I agree with you, I'd prefer for this to be outside of the class entirely.
I'll address this in this PR soon.

Malvoz commented on packages/hint-no-disallowed-headers/src/_locales/en/messages.json:
```diff
@@ -11,6 +11,30 @@
+    },
+    "disallowedXFrameOptionsHeader": {
+        "description": "Report message when the response includes the X-Frame-Options header",
+        "message": "The 'X-Frame-Options' header should not be used, the same effect can be achieved with more consistent support using the 'Content-Security-Policy' header."
```
I suggest explicitly mentioning the [`frame-ancestors`](https://www.w3.org/TR/CSP/#directive-frame-ancestors) navigation directive here as it might save developers some time if they're unfamiliar with it.

(Also, while I'm not sure there's much benefit in describing it here, `frame-ancestors` and `X-Frame-Options` doesn't have the _exact_ same effect, `frame-ancestors` is stronger in that it checks against _any_ ancestors where `X-Frame-Options` only checks the top-level document's location as described here: https://www.w3.org/TR/CSP/#frame-ancestors-and-frame-options)

captainbrosset replied:
Thanks, it makes sense to me to mention the `frame-ancestors` directive. I'll update the message.

captainbrosset commented:
Update: I've addressed all remaining review comments, and squashed the whole thing down to 1 commit.
I guess it's ready for a last round of reviews and/or merge whenever CI has completed.

antross commented:
Thanks @captainbrosset, all the updates look good to me!

@molant, @johnemau, @utsavized - do any of you have time to take a look?
